### PR TITLE
Implement InputList component

### PR DESCRIPTION
**Pull request by drborges**
### What does this PR doe?

Provide an `InputList` component that can manage lists of input fields providing add/remove functionality out-of-the-box.

Here're a few facts about the component:

1. Parent components may have access to the underlying data by providing an `onChange` event handler to the `InputList`.
2. The input field template is wrapped by `InputList.Input`, that tells the `InputList` how to render the input fields. The wrapped input will automatically receive from the `InputList` a `key`, `value`, `onRemove` and `onChange` props, which are used internally to keep the input field state up-to-date and to respond to item removal when `onRemove` is called.
3. The developer is free to style/structure the list rows/inputs as needed since no assumptions in that regard are made by the component.
4. Components wrapped by `InputList.Add` will receive an `onClick` prop which triggers adding new empty items to the input list.

### Example

```jsx
const incomeSources= [
  {id: 1, source: "Power", income: 50000},
  {id: 2, source: "Retirement Funds", income: 30000},
]

const IncomeSourceInput = ({ value, onChange, onRemove }) => (
  <div>
    <input type="checkbox" name="default" value={value.default} onChange={onChange} />
    <input name="source" value={value.source} onChange={onChange} />
    <input name="income" value={value.income} onChange={onChange} />
    <button type="button" onClick={onRemove}>{"Remove"}</button>
  </div>
)

const Form = ({ }) => (
  <InputList value={incomeSources} onChange=onChange()>
    <InputList.Add>
      <button type="button">Add</button>
    </InputList.Add>

    <InputList.Input>
      <IncomeSourceInput/>
    </InputList.Input>
  </InputList>
)
```

### CODE REVIEW

File: components/InputList/index.jsx [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    this.setState({
+      inputs: [
+        ...this.state.inputs.slice(0, i),
+        inputValue,
+        ...this.state.inputs.slice(i + 1)
+      ]
+    }, () => this.props.onChange(this.state.inputs))
+  }
+
+  handleAddClick = e => {
```
Comment (gmfvpereira) [inline components/InputList/index.jsx]: Do you need to declare this unused var?

**Comment by wadewinningham** [comment]
@gmfvpereira Why `nitro_react` rather than `nitro_storybook`?

**Comment by gmfvpereira** [comment]
@xjunior @rafbgarcia are also interested in this discussion ^

**Comment by thatjsguy** [comment]
@wadewinningham this needs to be discussed in an RFC to determine exactly what we should do *consistently* moving forward. The pain of local development against nitro-storybook along with the added pain of PR making (with all the versioning) is well.. a pain.

**Comment by xjunior** [comment]
@wadewinningham @gmfvpereira @drborges  @thatjsguy we got to the agreement that the storybook should be merely that, a place to try out components. Ideally we should prep 2 PR's to move the components into nitro_react.

**Comment by xjunior** [comment]
@thatjsguy I really don't think we need a PR since it's a stablished process. We once thought the storybook would be used for one thing and turned out we were wrong. The components produced on that period should be moved into `NitroReact`.

**Comment by thatjsguy** [comment]
@xjunior - I don't recall that exact agreement. In fact, we had discussed storybook as being the place where you *begin* creating stateless components. Regardless, this should be discussed officially - outside of this PR.

**Comment by wadewinningham** [comment]
As the main group involved in ironing out a decision on this are now involved in this PR, and since Create is just a week away, maybe we can discuss this over beers and if there's an RFC involved, we can use that discussion as a starting point for that.

File: components/InputList/index.jsx (hunk 1 above)
Comment (drborges) [reply components/InputList/index.jsx]: ah, not really :) I'll get rid of it.

**Comment by xjunior** [comment]
@thatjsguy I think that fits into "try out components". But the point is that the Storybook should be a storybook of the components we have implemented, but the library who implements that is nitro_react. For a perspective, when you install `react-bootstrap` you don't also get a storybook of it. But I think we're both agreeing on it :)

See you at create!

**Comment by thatjsguy** [comment]
@xjunior Yeh we are def on the same page here. Just need to find a way to smooth things out process wise and not leave UX out of the equation since they rely heavily on storybook.

**Comment by jasoncypret** [comment]
@xjunior @gmfvpereira "what I got from the storybook idea is that this should be like a sandbox where it's gonna be faster to implement a new reusable component, but not the repository of such components." && "we got to the agreement that the storybook should be merely that, a place to try out components. Ideally we should prep 2 PR's to move the components into nitro_react."

I don't know where you got this agreement, but this is false. In your case, Nova decided to "Delay" adding to storybook until the end of the project to speed up weekly sprint development. So your project with TA has to bake in moving that work at the end. I've spoken to Steve and Tim about this and they are aware that we need to do this work. It's important to make those components your team created available EVERYWHERE.

This repo is part of a move to develop out our design system for reuse. Have tighter control of UI, and make it easier to stay consistent with UI, document it, enable discovery, and see components side by side. This is public, we can use it for recruiting, and showing the outside world what power is about.

This repo is used also in 3 other applications, when we don't add things to Storybook we are not making those components reusable to everyone. We are not documenting how to use it. We are not making it easy to audit the UI for consistency. Recently we are considering moving Connect into its own repo. Storybook will need to be included there as well.

If there are implementation details we need to iron out, let's do that. But speaking with Tim last week, this is still very important for the business.

**Comment by rafbgarcia** [comment]
@jasoncypret I have a question:

This week I had an issue with the `DatePicker` component in nitro-storybook, the HTML structure it currently has is breaking its layout in the recruiting area of Nitro.

In nitro-react we also had a very similar unused `DatePicker` component, so I ended up using that one to fix Nitro's problem. That component is now fixed, should I move my fix to the DatePicker in nitro-storybook?

If so, how should we go for testing? Do we need to test everywhere where the component is being used?

I'm asking that because I believe that at some point the DatePicker component in nitro-storybook was updated to fix one problem in Nitro and caused it to be broken in other areas, for that same reason I asked Stephen Boligitz to test all places in Nitro where it's being used, but I don't think that'll be scalable as we move forward.

Maybe that happened because we're still not 100% consistent on the HTML we're using?

**Comment by drborges** [comment]
uhuu, a bit late to the party here... here's my 2cents:

It seems to me that it is a common agreement that the current storybook workflow (when working with Nitro in particular) is not *ideal* and introduces **a lot** of complexity and overhead with having to manage dep versions (`yarn clientx` + `./install.sh` + `yarn clientw` this combo takes **forever** to run). However, based on the big picture that @jasoncypret presented, I agree that reusability is the ultimate goal. Any new application in Power should be able to leverage this components system and thus allow people to prototype ideas quickly, with as little friction as possible.

With that said, it seems to me that we should focus on improving the workflow itself so we may leverage storybook for faster development feedback loop without letting Nitro ecosystem hit our productivity.

Here're some ideas:

1. Start the component development within storybook with **zero** assumptions of which application will be using it (Nitro, Standalone Connect, etc...). At this point, the developer can fully leverage storybook features (especially hot module reload) without being affected by the host app's ecosystem (Nitro and everything that comes along with it...).
2. Once you feel you're done with the component and it's time to test its integration with the host app (Nitro most often), use yarn link to point Nitro to your local version of storybook so you don't have to prematurely deal with Github tags/releases:

`package.json`

```diff
"dependencies": {
    "@fortawesome/react-fontawesome": "0.0.17",
    "@powerhome/js_test_helpers": "link:../js_test_helpers",
    "@powerhome/nitro_babel": "link:../nitro_babel",
+    "nitro-storybook": "link:../../../nitro-storybook",
}
```
3. Run `yarn clientx` + `./install.sh` + `yarn clientw` in `nitro-web` to make sure you are now using your local version of storybook.
4. Test the component integration, success?
5. Merge your storybook PR
6. Create a new Github release for the latest master revision
7. Remove the yarn link from `nitro-web` and make sure it points to the newly created release
8. Ship it.

Note that all the workflow related complexity starts when we introduce Nitro into the mix, however, if you can make sure that when you get to the point of testing your component within Nitro the work on the component is either wrapped up or close to being finished, the integration step with Nitro won't hurt as much as it would if you were to develop the component while integrating it with Nitro.

This is obviously a recomendation on how to reduce the integration friction with Nitro and not necessarily the ultimate solution to the problem.

To me, the root source of the workflow problem here is likely in the tooling, if wiping out yarn's cache and installing a new dependency (new storybook version for instance) were simpler, faster, probably working off of storybook wouldn't be issue in the first place.

**Comment by wadewinningham** [comment]
@jasoncypret @rafbgarcia Based on the previous issue that Rafa pointed out, I'd be inclined to have a Ninja (or the dev) check out other areas of Nitro using the same component.

Generally though, I would think most components would be relatively self-sustaining visually and usability-wise. The only thing I'd *expect* could cause issues would be incoming *props*. But that would just expose an internal issue to the control anyway.

**Comment by thatjsguy** [comment]
RFC is here https://github.com/powerhome/rfcs/pull/47 for discussion purposes

**Comment by drborges** [comment]
BTW, since we were not able to leverage this component for its original purpose on nitro-web, I'll be closing this PR for now, if anyone sees a use case for this, feel free to reopen this PR, I just don't want to ship code that isn't being used anywhere.

**Pull request closed without merging**